From stable-bounces@linux.kernel.org  Fri Feb 24 13:09:14 2006
Date: Fri, 24 Feb 2006 13:04:24 -0800
From: akpm@osdl.org
To: torvalds@osdl.org
Cc: j-nomura@ce.jp.nec.com, stable@kernel.org, agk@redhat.com
Subject: [PATCH] dm: missing bdput/thaw_bdev at removal


From: "Jun'ichi Nomura" <j-nomura@ce.jp.nec.com>

Need to unfreeze and release bdev otherwise the bdev inode with
inconsistent state is reused later and cause problem.

Signed-off-by: Jun'ichi Nomura <j-nomura@ce.jp.nec.com>
Acked-by: Alasdair G Kergon <agk@redhat.com>
Cc: <stable@kernel.org>
Signed-off-by: Andrew Morton <akpm@osdl.org>
[chrisw: backport to 2.6.15]
Signed-off-by: Chris Wright <chrisw@sous-sol.org>
---

 drivers/md/dm.c |    4 ++++
 1 files changed, 4 insertions(+)

--- linux-2.6.15.4.orig/drivers/md/dm.c
+++ linux-2.6.15.4/drivers/md/dm.c
@@ -812,6 +812,10 @@ static struct mapped_device *alloc_dev(u
 
 static void free_dev(struct mapped_device *md)
 {
+	if (md->frozen_bdev) {
+		thaw_bdev(md->frozen_bdev, NULL);
+		bdput(md->frozen_bdev);
+	}
 	free_minor(md->disk->first_minor);
 	mempool_destroy(md->tio_pool);
 	mempool_destroy(md->io_pool);
